### PR TITLE
Building `libicuc` from source

### PR DESCRIPTION
**Pull request by soumyaDghosh**
`icu73` is needed to build `webkit2gtk-5` from source. That's why here I added the `icu73` first. We need the `webkit2gtk5` here soon, cause my gtk apps use it now.

### CODE REVIEW

File: snapcraft.yaml [hunk 1]
```diff
@@ -1382,9 +1381,20 @@ parts:
...
-
+  libicuc:
+    source: https://github.com/unicode-org/icu.git
+    source-tag: release-73-1
+    source-subdir: icu4c/source
+    plugin: autotools
+    autotools-configure-parameters:
+      # WORKAROUND: Install to /usr instead of /usr/local because it's not in search paths
+      - --prefix=/usr
+      - --disable-renaming
```
Comment (sergio-costas) [inline snapcraft.yaml:1396]: Can you test to add `- --libdir=\${exec_prefix}/lib/$CRAFT_ARCH_TRIPLET` to install libraries into /usr/lib/x86_64-... and check that it does work as expected?

File: snapcraft.yaml [hunk 2]
```diff
@@ -1382,9 +1381,20 @@ parts:
         sed -i 's#libdir=/usr#libdir=${prefix}#' $PC
         sed -i 's#includedir=/usr#includedir=${prefix}#' $PC
       done
-
+  libicuc:
+    source: https://github.com/unicode-org/icu.git
+    source-tag: release-73-1
```
Comment (sergio-costas) [inline snapcraft.yaml:1387]: Can you add major-minor version data?
```
# ext:updatesnap
#   version-format:
#     format: 'release-%M-%m'
```

File: snapcraft.yaml (hunk 1 above)
Comment (soumyaDghosh) [reply snapcraft.yaml:1396]: Actually I was getting this error while building it, but I guess this will fix it.
```
2023-05-02T16:28:02.2492240Z :: Extracting symbols from /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Compress/Raw/Zlib/Zlib.so into usr/lib/debug/.build-id/7b/ef161cb7a5d54796c8e6028ac11322ab3ce77f.debug
2023-05-02T16:28:02.2501497Z :: Stripping with --strip-unneeded  /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Compress/Raw/Zlib/Zlib.so
2023-05-02T16:28:02.2510312Z :: Extracting symbols from /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Math/BigInt/FastCalc/FastCalc.so into usr/lib/debug/.build-id/5b/2317dbb6d1135167696d33ef8d4c44740b3d46.debug
2023-05-02T16:28:02.2517444Z :: Stripping with --strip-unneeded  /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Math/BigInt/FastCalc/FastCalc.so
2023-05-02T16:28:02.2525001Z :: Extracting symbols from /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Filter/Util/Call/Call.so into usr/lib/debug/.build-id/d0/1cfc939a6f47ccde67102d04897f728bd212cb.debug
2023-05-02T16:28:02.2531375Z :: Stripping with --strip-unneeded  /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Filter/Util/Call/Call.so
2023-05-02T16:28:02.2538354Z :: Extracting symbols from /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Hash/Util/FieldHash/FieldHash.so into usr/lib/debug/.build-id/77/ca84b68b81a6c4d8434953d58dae2805a69750.debug
2023-05-02T16:28:02.2544160Z :: Stripping with --strip-unneeded  /root/prime/usr/lib/x86_64-linux-gnu/perl/5.34.0/auto/Hash/Util/FieldHash/FieldHash.so
2023-05-02T16:28:02.2554149Z :: Traceback (most recent call last):
2023-05-02T16:28:02.2554893Z ::   File "/usr/local/lib/python3.10/dist-packages/craft_parts/sources/git_source.py", line 78, in generate_version
2023-05-02T16:28:02.2566662Z ::     subprocess.check_output(
2023-05-02T16:28:02.2567044Z ::   File "/usr/lib/python3.10/subprocess.py", line 420, in check_output
2023-05-02T16:28:02.2567480Z ::     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
2023-05-02T16:28:02.2567891Z ::   File "/usr/lib/python3.10/subprocess.py", line 524, in run
2023-05-02T16:28:02.2568271Z ::     raise CalledProcessError(retcode, process.args,
2023-05-02T16:28:02.2580916Z :: subprocess.CalledProcessError: Command '['git', '-C', '/data', 'describe', '--dirty']' returned non-zero exit status 128.
2023-05-02T16:28:02.2581351Z ::
2023-05-02T16:28:02.2582172Z :: The above exception was the direct cause of the following exception:
2023-05-02T16:28:02.2582471Z ::
2023-05-02T16:28:02.2582722Z :: Traceback (most recent call last):
2023-05-02T16:28:02.2590350Z ::   File "/data/strip_binaries.py", line 95, in <module>
2023-05-02T16:28:02.2591033Z ::     version_number = GitSource.generate_version(part_src_dir=os.environ['CRAFT_PROJECT_DIR'])
2023-05-02T16:28:02.2591775Z ::   File "/usr/local/lib/python3.10/dist-packages/craft_parts/sources/git_source.py", line 97, in generate_version
2023-05-02T16:28:02.2599412Z ::     raise errors.VCSError(message=stderr.decode(encoding).strip()) from err
2023-05-02T16:28:02.2600182Z :: craft_parts.sources.errors.VCSError: fatal: detected dubious ownership in repository at '/data'
2023-05-02T16:28:02.2600620Z :: To add an exception for this directory, call:
2023-05-02T16:28:02.2607463Z ::
2023-05-02T16:28:02.2607951Z :: 	git config --global --add safe.directory /data
2023-05-02T16:28:02.4308516Z 'override-prime' in part 'cleanup' failed with code 1.
2023-05-02T16:28:02.4309426Z Review the scriptlet and make sure it's correct.
```
I first thought, that may be due to building this in github made this issue.

File: snapcraft.yaml (hunk 2 above)
Comment (soumyaDghosh) [reply snapcraft.yaml:1387]: I am not able to understand this.. can you explain me this a little bit more?
should it be like this?
```
# ext:updatesnap
#   version-format:
#     format: 'release-%M-%m'
```
Major 73 and minor 1?

Comment (sergio-costas) [reply snapcraft.yaml:1387]: Ops... sorry, I have to add some documentation about this. We use internally some tools to automagically track new versions of each part; but since each project uses its own format for the tags, we need a way of specifying how to parse them. By default, if the format is "MAJOR.MINOR.REVISION" or "vMAJOR.MINOR.REVISION", it can be managed automatically; but for more exotic formats, like in this case, we have to specify how to parse the tags. In this case, that says that the tags are in "release-MAJOR-MINOR" format.

Also, since snapcraft would bark if we just add those extra elements, we add them as comments, and our tool searches for a comment that begins with `# ext:updatesnap`, and replaces the '#' in the next lines with a blank space, thus "transforming" that into standard YAML code that can be then processed by us.

File: snapcraft.yaml (hunk 1 above)
Comment (sergio-costas) [reply snapcraft.yaml:1396]: Wow... that looks like a bug in the debug symbols extracting tool...

Comment (sergio-costas) [reply snapcraft.yaml:1396]: Mmm... try running `git config --global --add safe.directory /data` in your folder...

Comment (soumyaDghosh) [reply snapcraft.yaml:1396]: Still I am building these on github action... I can't.... How can I test them in a local machine😭😭😭😭, It'll take days. Let me see what happens, now.

Comment (sergio-costas) [reply snapcraft.yaml:1396]: Oh... are you building it inside github????

You should be able to build it using "snapcraft", as usual, in your local machine. But it takes about half an hour or so...

Comment (sergio-costas) [reply snapcraft.yaml:1396]: Or am I missing something?

Comment (soumyaDghosh) [reply snapcraft.yaml:1396]: > Or am I missing something?

With my 10Mb/s speed, it takes me hours

Comment (sergio-costas) [reply snapcraft.yaml:1396]: It's not only the 10Mb/s; downloading is only a little part of the work. Compiling all those parts use A LOT of time, unfortunately. It would happen to you even with a 1Gb/s link, I'm afraid...

Comment (soumyaDghosh) [reply snapcraft.yaml:1396]: I got it built up. I added the `git config --global --add safe.directory /data` before the python line. Check the actions, you'll get artifact.
https://github.com/soumyaDghosh/gnome-sdk/actions/runs/4881611247

File: snapcraft.yaml (hunk 2 above)
Comment (soumyaDghosh) [reply snapcraft.yaml:1387]: I have added this also.

File: .github/workflows/build-prs.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+name: Build Pull Requests
```
Comment (sergio-costas) [inline .github/workflows/build-prs.yml]: What is this for...?

Comment (sergio-costas) [reply .github/workflows/build-prs.yml]: I mean... if it's not needed anymore, it's better to remove it.

File: snapcraft.yaml [hunk 4]
```diff
@@ -1633,6 +1647,7 @@ parts:
 
       OLDPATH=$PATH
       python3 -m pip install pyelftools craft-parts==1.14.0
+      git config --global --add safe.directory /data
```
Comment (sergio-costas) [inline snapcraft.yaml]: The point is... what is this "data" folder?

Comment (sergio-costas) [reply snapcraft.yaml]: Also, it doesn't seem that it has been needed until now... Are you sure that it's not an artifact of your system?

Comment (soumyaDghosh) [reply snapcraft.yaml]: Okay, wait.. let me remove the github action files and try to build this locally

Comment (soumyaDghosh) [reply snapcraft.yaml]: Yes, tested locally.. no problem, you can give that a try too

File: snapcraft.yaml [hunk 5]
```diff
@@ -33,6 +33,7 @@ parts:
       - GDK_PIXBUF_MODULE_FILE: $CRAFT_STAGE/usr/lib/$CRAFT_ARCH_TRIPLET/gdk-pixbuf-2.0/2.10.0/loaders.cache
       - PKG_CONFIG_PATH: $CRAFT_STAGE/usr/lib/$CRAFT_ARCH_TRIPLET/pkgconfig:$CRAFT_STAGE/usr/lib/pkgconfig:$CRAFT_STAGE/usr/share/pkgconfig${PKG_CONFIG_PATH:+:$PKG_CONFIG_PATH}
       - CRAFT_EXT_CORE_LEVEL: core22
+      - PYTHONPATH: $PYTHONPATH:$CRAFT_STAGE/lib/python3.10/site-packages
```
Comment (sergio-costas) [inline snapcraft.yaml:36]: Why is this needed? Is there a problem with the current pythonpath?

File: snapcraft.yaml [hunk 6]
```diff
@@ -705,7 +706,6 @@ parts:
     override-pull: |
       craftctl default
       patch -p1 < $CRAFT_PROJECT_DIR/patches/style-manager-Support-Yaru-accent-colors.patch
-
```
Comment (sergio-costas) [inline snapcraft.yaml:708]: Please, keep separation between parts for clarity.

File: snapcraft.yaml [hunk 7]
```diff
@@ -1382,9 +1382,105 @@ parts:
         sed -i 's#libdir=/usr#libdir=${prefix}#' $PC
         sed -i 's#includedir=/usr#includedir=${prefix}#' $PC
       done
-
```
Comment (sergio-costas) [inline snapcraft.yaml:1385]: Idem.

File: snapcraft.yaml [hunk 8]
```diff
@@ -1382,9 +1382,105 @@ parts:
...
+      - -bin/python3
+      - -bin/python3.10
+      - -bin/pip
+      - -bin/pip3
+      - -bin/pip3.10
+      - -pyvenv.cfg  
+  webkitgtk:
+    after: [gi-docgen, libicuc]
+    source: https://salsa.debian.org/webkit-team/webkit.git
+    source-tag: upstream/2.40.1
```
Comment (sergio-costas) [inline snapcraft.yaml:1424]: Mmm... but webkitgtk 2.40 is already being included...

image

Comment (sergio-costas) [reply snapcraft.yaml:1424]: Or are you just adding this to "jump" to version 2.50 as soon as it goes out?

**Comment by soumyaDghosh** [comment]
First of all I'm extremely sorry about these.. These were intended for testing. But, I forgot that the PR is still open.

**Comment by soumyaDghosh** [comment]
I am closing this. This has got very polluted.
